### PR TITLE
315 activity publish backend new

### PR DESCRIPTION
**Pull request by Sanilblank**

### CODE REVIEW

File: app/Http/Controllers/Admin/Workflow/ActivityWorkflowController.php [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Publish an activity.
+     *
+     * @param $id
+     *
+     * @return JsonResponse|RedirectResponse
+     */
+    public function publish($id): JsonResponse|RedirectResponse
```
Comment (sriza) [inline app/Http/Controllers/Admin/Workflow/ActivityWorkflowController.php:49]: only JsonResponse is returned by the function

File: app/Http/Controllers/Admin/Workflow/ActivityWorkflowController.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Validates activity on the IATI Validator.
+     *
+     * @param $id
+     *
+     * @return JsonResponse|RedirectResponse
+     */
+    public function validateActivity($id): JsonResponse|RedirectResponse
```
Comment (sriza) [inline app/Http/Controllers/Admin/Workflow/ActivityWorkflowController.php:138]: only JsonResponse is returned

File: app/IATI/Repositories/Activity/ActivityRepository.php [hunk 3]
```diff
@@ -58,4 +58,47 @@ public function getActivityForOrganization($organizationId, int $page = 1): Coll
...
+    }
+
+    /**
+     * Sets activity status to draft.
+     *
+     * @param $activity_id
+     *
+     * @return null
+     */
+    public function resetActivityWorkflow($activity_id)
```
Comment (sriza) [inline app/IATI/Repositories/Activity/ActivityRepository.php:100]: return type of function

File: app/IATI/Services/Activity/ActivityService.php [hunk 4]
```diff
@@ -104,6 +105,72 @@ public function getActivity($id): Activity
...
+    }
+
+    /**
+     * Returns required service file.
+     *
+     * @param $serviceName
+     *
+     * @return \Illuminate\Contracts\Foundation\Application|mixed
+     */
+    public function getService($serviceName)
```
Comment (sriza) [inline app/IATI/Services/Activity/ActivityService.php:130]: return type of function

File: app/IATI/Services/Activity/ActivityService.php [hunk 5]
```diff
@@ -104,6 +105,72 @@ public function getActivity($id): Activity
...
+    }
+
+    /**
+     * Sets activity status to draft.
+     *
+     * @param $activity_id
+     *
+     * @return null
+     */
+    public function resetActivityWorkflow($activity_id)
```
Comment (sriza) [inline app/IATI/Services/Activity/ActivityService.php:169]: return type of function missing

File: app/IATI/Services/Activity/ConditionService.php [hunk 6]
```diff
@@ -88,4 +93,55 @@ public function formGenerator($id): Form
...
     }
+
+    /**
+     * Returns data in required xml array format.
+     *
+     * @param Activity $activity
+     *
+     * @return array
+     */
+    public function getXmlData(Activity $activity)
```
Comment (sriza) [inline app/IATI/Services/Activity/ConditionService.php:104]: return type missing

File: app/IATI/Services/Publisher/PublisherService.php [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+            }
+
+            throw  $exception;
+        }
+    }
+
+    /**
+     * Updates activity published table.
+     */
+    protected function updateStatus()
```
Comment (sriza) [inline app/IATI/Services/Publisher/PublisherService.php:247]: void return type missing

File: app/IATI/Services/Workflow/ActivityWorkflowService.php [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    /**
+     * Publish an activity to the IATI registry.
+     *
+     * @param $activity
+     *
+     * @return void
+     */
+    public function publishActivity($activity)
```
Comment (sriza) [inline app/IATI/Services/Workflow/ActivityWorkflowService.php:98]: add return type to function

File: app/IATI/Services/Workflow/ActivityWorkflowService.php [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    /**
+     * Unpublish activity from the IATI registry.
+     *
+     * @param $activity
+     *
+     * @return void
+     */
+    public function unpublishActivity($activity)
```
Comment (sriza) [inline app/IATI/Services/Workflow/ActivityWorkflowService.php:128]: return type missing

File: app/IATI/Services/Workflow/ActivityWorkflowService.php [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * Removes activity file name from activity published row.
+     *
+     * @param $publishedFile
+     * @param $activity
+     *
+     * @return void
+     */
+    public function removeActivityFromPublishedArray($publishedFile, $activity)
```
Comment (sriza) [inline app/IATI/Services/Workflow/ActivityWorkflowService.php:144]: return type missing

File: app/IATI/Services/Workflow/RegistryApiHandler.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Initialize an CkanClient instance.
+     *
+     * @param $url
+     * @param $key
+     *
+     * @return RegistryApiHandler
+     */
+    public function init($url, $key)
```
Comment (sriza) [inline app/IATI/Services/Workflow/RegistryApiHandler.php:35]: return type missing

File: app/IATI/Services/Workflow/RegistryApiHandler.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        $this->client = new CkanClient($url, $key);
+        $this->apiKey = $key;
+
+        return $this;
+    }
+
+    /**
+     * @param $publisherId
+     */
+    public function setPublisher($publisherId)
```
Comment (sriza) [inline app/IATI/Services/Workflow/RegistryApiHandler.php:46]: return type missing as well as defination of function

File: app/IATI/Services/Xml/XmlGeneratorService.php [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+     * @param $activity
+     * @param $transaction
+     * @param $result
+     * @param $settings
+     * @param $organization
+     *
+     * @return void
+     */
+    public function generateActivityXml($activity, $transaction, $result, $settings, $organization)
+    {
```
Comment (sriza) [inline app/IATI/Services/Xml/XmlGeneratorService.php:51]: return type missing

File: app/IATI/Services/Xml/XmlGeneratorService.php [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * Generates new xml file after unpublishing.
+     *
+     * @param $publishedFile
+     *
+     * @return void
+     */
+    public function generateNewXmlFile($publishedFile)
```
Comment (sriza) [inline app/IATI/Services/Xml/XmlGeneratorService.php:74]: return type missing

File: app/IATI/Services/Xml/XmlGeneratorService.php [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * Deletes the unpublished file from server.
+     *
+     * @param $filename
+     *
+     * @return void
+     */
+    public function deleteUnpublishedFile($filename)
```
Comment (sriza) [inline app/IATI/Services/Xml/XmlGeneratorService.php:62]: return type missing

File: app/IATI/Services/Xml/XmlSchemaErrorParser.php [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+    /**
+     * get modified message send as per the schema.
+     *
+     * @param $error
+     * @param $validateXml
+     *
+     * @return mixed|string
+     */
+    public function getModifiedError($error, $validateXml)
```
Comment (sriza) [inline app/IATI/Services/Xml/XmlSchemaErrorParser.php:30]: return type missing in all the functions of this file

File: app/IATI/Traits/RegistryApiInvoker.php [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Make an api request to the given action.
+     *
+     * @param        $action
+     * @param        $requestParameter
+     * @param null $apiKey
+     *
+     * @return mixed
+     */
+    protected function request($action, $requestParameter = null, $apiKey = null)
```
Comment (sriza) [inline app/IATI/Traits/RegistryApiInvoker.php:23]: return type missing in all the functions of this file

File: app/IATI/Traits/XmlServiceTrait.php [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+trait XmlServiceTrait
+{
+    /**
+     * return xml validation message with type.
+     *
+     * @param $error
+     *
+     * @return string
+     */
+    protected function libxml_display_error($error)
```
Comment (sriza) [inline app/IATI/Traits/XmlServiceTrait.php:17]: return type missing in all the function of this file

File: app/Observers/ActivityObserver.php [hunk 19]
```diff
@@ -91,6 +92,19 @@ public function created(Activity $activity): void
...
     }
+
+    /**
+     * Resets activity status to draft.
+     *
+     * @param $model
+     *
+     * @return void
+     */
+    public function resetActivityStatus($model)
```
Comment (sriza) [inline app/Observers/ActivityObserver.php:106]: return type missing

File: app/Observers/ResultObserver.php [hunk 20]
```diff
@@ -68,5 +69,20 @@ public function created(Result $result): void
...
+    }
+
+    /**
+     * Resets activity status to draft.
+     *
+     * @param $result
+     *
+     * @return void
+     */
+    public function resetActivityStatus($result)
```
Comment (sriza) [inline app/Observers/ResultObserver.php:82]: return type missing

File: app/Observers/TransactionObserver.php [hunk 21]
```diff
@@ -67,5 +68,20 @@ public function created(Transaction $transaction): void
...
+    }
+
+    /**
+     * Resets activity status to draft.
+     *
+     * @param $transaction
+     *
+     * @return void
+     */
+    public function resetActivityStatus($transaction)
```
Comment (sriza) [inline app/Observers/TransactionObserver.php:81]: return type missing

File: app/IATI/Services/Workflow/ActivityWorkflowService.php (hunk 8 above)
Comment (Sanilblank) [reply app/IATI/Services/Workflow/ActivityWorkflowService.php:98]: This function returns void

File: app/IATI/Services/Workflow/ActivityWorkflowService.php (hunk 9 above)
Comment (Sanilblank) [reply app/IATI/Services/Workflow/ActivityWorkflowService.php:128]: Function returns void

File: app/Http/Controllers/Admin/Activity/ActivityController.php [hunk 22]
```diff
@@ -136,10 +151,17 @@ public function show($id): View|JsonResponse|RedirectResponse
             $transactions = $this->transactionService->getActivityTransactions($activity->id);
             $status = $activity->element_status;
             $progress = $this->activityService->activityPublishingProgress($activity);
+            $coreCompleted = isCoreElementCompleted(array_merge(['reporting_org' => $activity->organization->reporting_org_complete_status], $activity->element_status));
+            $iatiValidatorResponse = null;
```
Comment (man-git007) [inline app/Http/Controllers/Admin/Activity/ActivityController.php:155]: use ternary

$validatorResponse = $this->activityValidatorResponseService->getValidatorResponse($id);
$iatiValidatorResponse = ($validatorResponse)?$validatorResponse->response:null;

File: app/Http/Controllers/Admin/Activity/ResultController.php [hunk 23]
```diff
@@ -161,7 +166,11 @@ public function edit($activityId, $resultId): View|RedirectResponse
             $element = json_decode(file_get_contents(app_path('IATI/Data/elementJsonSchema.json')), true);
```
Comment (man-git007) [inline app/Http/Controllers/Admin/Activity/ResultController.php:166]: use `getElementSchema('result')` function

File: app/Http/Requests/Activity/OtherIdentifier/OtherIdentifierRequest.php [hunk 24]
```diff
@@ -68,9 +65,57 @@ public function getMessagesForOtherIdentifier(array $formFields): array
...
+
+    /**
+     * Return rules for owner org.
+     *
+     * @param $formFields
+     * @param $formBase
+     *
+     * @return array
+     */
+    public function getRulesForOwnerOrg($formFields, $formBase)
```
Comment (man-git007) [inline app/Http/Requests/Activity/OtherIdentifier/OtherIdentifierRequest.php:88]: add return type `:array`

File: app/IATI/Elements/Xml/ArrayToXml.php [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+    private static $encoding = 'UTF-8';
+
+    /**
+     * Initialize the root XML node [optional].
+     *
+     * @param $version
+     * @param $encoding
+     * @param $format_output
+     */
+    public static function init($version = '1.0', $encoding = 'UTF-8', $format_output = true)
```
Comment (man-git007) [inline app/IATI/Elements/Xml/ArrayToXml.php:28]: add return type `void` in php doc and function

File: app/IATI/Elements/Xml/ArrayToXml.php [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * Convert an Array to XML.
+     *
+     * @param string $node_name - name of the root node to be converted
+     * @param array  $arr       - aray to be converted
+     *
+     * @return \DomDocument
+     */
+    public static function &createXML($node_name, $arr = [])
```
Comment (man-git007) [inline app/IATI/Elements/Xml/ArrayToXml.php:43]: add return type `:DomDocument`

File: app/IATI/Models/Activity/ActivityPublished.php [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+     * Returns filename.
+     *
+     * @param $delimiter
+     * @param $file
+     *
+     * @return string
+     */
+    protected function getFilename($delimiter, $file): string
+    {
+        $file = explode($delimiter, $file);
```
Comment (man-git007) [inline app/IATI/Models/Activity/ActivityPublished.php:79]: return reset(explode($delimiter, $file));

File: app/IATI/Repositories/Activity/ActivityPublishedRepository.php [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace App\IATI\Repositories\Activity;
+
+use App\IATI\Models\Activity\ActivityPublished;
+use Illuminate\Database\Eloquent\Model;
+
+/**
+ * Class ActivityPublishedRepository.
+ */
+class ActivityPublishedRepository
```
Comment (man-git007) [inline app/IATI/Repositories/Activity/ActivityPublishedRepository.php:13]: extend `Repository`

File: app/IATI/Repositories/Activity/ActivitySnapshotRepository.php [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+declare(strict_types=1);
+
+namespace App\IATI\Repositories\Activity;
+
+use App\IATI\Models\Activity\ActivitySnapshot;
+
+/**
+ * Class ActivitySnapshotRepository.
+ */
+class ActivitySnapshotRepository
```
Comment (man-git007) [inline app/IATI/Repositories/Activity/ActivitySnapshotRepository.php:12]: Extend `Repository`

File: app/IATI/Repositories/Validator/ActivityValidatorResponseRepository.php [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * Deletes the validator response of the unpublished activity.
+     *
+     * @param $activityId
+     *
+     * @return bool
+     */
+    public function deleteValidatorResponse($activityId): bool
+    {
+        $validatorResponse = $this->model->where('activity_id', $activityId)->first();
```
Comment (man-git007) [inline app/IATI/Repositories/Validator/ActivityValidatorResponseRepository.php:63]: delete function of Repository can be used

File: app/IATI/Services/Activity/ActivityService.php [hunk 31]
```diff
@@ -104,6 +112,72 @@ public function getActivity($id): Activity
         return $this->activityRepository->find($id);
     }
 
+    /**
+     * Generates toast array.
+     *
+     * @return array
+     */
+    public function generateToastData(): array
```
Comment (man-git007) [inline app/IATI/Services/Activity/ActivityService.php:120]: move function to general.php file

File: app/IATI/Services/Publisher/PublisherService.php [hunk 32]
```diff
@@ -0,0 +1,259 @@
...
+        $this->searchForPublisher($this->publisherId);
+        $this->publishToRegistry($organization, $activityPublished->filename);
+    }
+
+    /**
+     * Set the file attribute.
+     *
+     * @param $activityPublished
+     */
+    protected function setFile($activityPublished)
```
Comment (man-git007) [inline app/IATI/Services/Publisher/PublisherService.php:49]: add return type `void`

File: app/IATI/Services/Publisher/PublisherService.php [hunk 33]
```diff
@@ -0,0 +1,259 @@
...
+     * Publish File to the IATI Registry.
+     *
+     * @param $organization
+     * @param $filename
+     *
+     * @return void
+     */
+    protected function publishToRegistry($organization, $filename): void
+    {
+        $data = $this->generatePayload($organization, $filename);
```
Comment (man-git007) [inline app/IATI/Services/Publisher/PublisherService.php:64]: $data = $this->generatePayload($organization, $filename);

    if ($this->isPackageAvailable($this->extractPackage($filename), $this->apiKey)) {
        $this->client->package_update($data);
    } else {
        $this->client->package_create($data);
    }

File: app/IATI/Services/Publisher/PublisherService.php [hunk 34]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * Get the data type or country code/region code from the filename.
+     *
+     * @param $filename
+     *
+     * @return string
+     */
+    protected function getCode($filename): string
+    {
+        $filename = str_replace('.xml', '', $filename);
```
Comment (man-git007) [inline app/IATI/Services/Publisher/PublisherService.php:105]: return substr(str_replace('.xml', '', $filename), strlen($this->publisherId) + 1);

File: app/IATI/Services/Publisher/PublisherService.php [hunk 35]
```diff
@@ -0,0 +1,259 @@
...
+     * Extract title for the file being published.
+     *
+     * @param $organization
+     * @param $fileType
+     *
+     * @return string
+     */
+    protected function extractTitle($organization, $fileType): string
+    {
+        if ($fileType == 'organisation') {
```
Comment (man-git007) [inline app/IATI/Services/Publisher/PublisherService.php:152]: use '==='

File: app/IATI/Services/Publisher/PublisherService.php [hunk 36]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * Extract the package name from the published filename.
+     *
+     * @param $filename
+     *
+     * @return string
+     */
+    protected function extractPackage($filename): string
+    {
+        $array = explode('.', $filename);
```
Comment (man-git007) [inline app/IATI/Services/Publisher/PublisherService.php:168]: return Arr::get(explode('.', $filename), 0, '');

**Review by man-git007 — COMMENTED** [review]
Codes reviewed

File: app/IATI/Services/Publisher/PublisherService.php [hunk 37]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    /**
+     * Updates activity published table.
+     *
+     * @return void
+     */
+    protected function updateStatus(): void
+    {
+        $this->activityPublished->published_to_registry = 1;
```
Comment (man-git007) [inline app/IATI/Services/Publisher/PublisherService.php:256]: Use repository pattern

File: app/IATI/Services/Validator/ActivityValidatorResponseService.php [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+
+declare(strict_types=1);
+
+namespace App\IATI\Services\Validator;
+
+use App\IATI\Repositories\Validator\ActivityValidatorResponseRepository;
+use Illuminate\Database\Eloquent\Model;
+
+/**
+ * Class ActivityValidatorService.
```
Comment (man-git007) [inline app/IATI/Services/Validator/ActivityValidatorResponseService.php:11]: in correct class name in php doc

File: app/IATI/Services/Xml/XmlSchemaErrorParser.php [hunk 39]
```diff
@@ -0,0 +1,153 @@
...
+     *
+     * @param $errorCodeElement
+     * @param $elementsInXml
+     * @param $xmlLines
+     * @param $errorLine
+     *
+     * @return mixed
+     */
+    protected function getMainElement($errorCodeElement, $elementsInXml, $xmlLines, $errorLine): mixed
+    {
```
Comment (man-git007) [inline app/IATI/Services/Xml/XmlSchemaErrorParser.php:110]: if (in_array($errorCodeElement, $elementsInXml)) {
    return $errorCodeElement;
}
$errorLine = $errorLine - 1;
$errorCodeElement = $this->getErrorElementName($xmlLines, $errorLine);

return $this->getMainElement($errorCodeElement, $elementsInXml, $xmlLines, $errorLine);

File: app/IATI/Traits/XmlBaseElement.php [hunk 40]
```diff
@@ -0,0 +1,174 @@
+<?php
+
+namespace App\IATI\Traits;
```
Comment (man-git007) [inline app/IATI/Traits/XmlBaseElement.php:3]: add `declare strict type`

File: app/IATI/Traits/XmlServiceTrait.php [hunk 41]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace App\IATI\Traits;
```
Comment (man-git007) [inline app/IATI/Traits/XmlServiceTrait.php:3]: add `declare strict type`

File: app/IATI/Traits/XmlServiceTrait.php [hunk 42]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /**
+     * return xml validation error messages.
+     *
+     * @return array
+     */
+    protected function libxml_fetch_errors(): array
+    {
+        $errors = libxml_get_errors();
```
Comment (man-git007) [inline app/IATI/Traits/XmlServiceTrait.php:46]: `return libxml_get_errors()`

**Review by man-git007 — COMMENTED** [review]
code review completed

File: app/IATI/Models/Activity/ActivityPublished.php (hunk 27 above)
Comment (Sanilblank) [reply app/IATI/Models/Activity/ActivityPublished.php:79]: Only variables can be passed by reference bhanera bhanxa dai yetikai pathauna mildaina raixa pass by ref garda

File: app/IATI/Repositories/Validator/ActivityValidatorResponseRepository.php (hunk 30 above)
Comment (Sanilblank) [reply app/IATI/Repositories/Validator/ActivityValidatorResponseRepository.php:63]: delete is being performed by using activity id which is not actually the id of the validator response table

**Pull request merged**